### PR TITLE
[LV][EVL] Emit vp.merge intrinsic to enable out-loop reduction in EVL vectorization.

### PR DESCRIPTION
**Pull request by Mel-Chen**
Following #90184, this patch emits vp.merge intrinsic, which is used to set the inactive lanes in a select operation to the RHS instead of undef. Currently, it is applied to out-loop reduction for EVL vectorization.
This patch performs transformation to convert 
```
  select(header_mask, LHS, RHS)
```
into
```
  vp.merge(all-true, LHS, RHS, EVL)
```  
And always use the predicated reduction select to set the incoming value of the reduction phi to support out-loop reduction when using tail folding with EVL. 

TODO: Postpone the adjustment of the predicated reduction select to VPlanTransform. The current adjustment might be too early, which could lead to a situation where the predicated reduction select is adjusted, but the EVL recipes cannot be successfully generated during VPlanTransform.

### CODE REVIEW

**Comment by Mel-Chen** [comment]
Move to draft since I found a bug about ReductionPHI. :[
Will re-open this PR after the bug is fixed.

**Comment by Mel-Chen** [comment]
538305be789acc7b2110d5bfa38636ec395e8d03 Fixed backedge of ReductionPHI.
Re-open this PR. Please take a look, thanks.

**Comment by Mel-Chen** [comment]
Ping

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp [hunk 1]
```diff
@@ -1383,6 +1385,16 @@ static void transformRecipestoEVLRecipes(VPlan &Plan, VPValue &EVL) {
                 VPValue *NewMask = GetNewMask(Red->getCondOp());
                 return new VPReductionEVLRecipe(*Red, EVL, NewMask);
               })
+              .Case<VPInstruction>([&](VPInstruction *VPI) -> VPRecipeBase * {
+                VPValue *LHS, *RHS;
+                if (!match(VPI, m_Select(m_Specific(HeaderMask), m_VPValue(LHS),
+                                         m_VPValue(RHS))))
+                  return nullptr;
+                VPValue *Cond = Plan.getOrAddLiveIn(ConstantInt::getTrue(Ctx));
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:1393]: Could you add a comment why true is used here? IIUC it is because it is limited to selects on the header mask for now?

Comment (Mel-Chen) [reply llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp]: Your understanding is absolutely correct. :)
And I can't find any other place where vp.merge is used.
7bfd0b20050a9d3859a2065c1b5622986d91f9c1

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 2]
```diff
@@ -661,7 +662,16 @@ Value *VPInstruction::generate(VPTransformState &State) {
     }
     return NewPhi;
   }
-
+  case VPInstruction::MergeUntilPivot: {
+    Value *Cond = State.get(getOperand(0), 0);
+    Value *OnTrue = State.get(getOperand(1), 0);
+    Value *OnFalse = State.get(getOperand(2), 0);
```
[the remark this answers is not in the document]
Comment (Mel-Chen) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp]: 6c2873b30497c15fd1ff197004da9acc485b3c2d

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 3]
```diff
@@ -661,7 +662,16 @@ Value *VPInstruction::generate(VPTransformState &State) {
...
+    Value *Pivot = State.get(getOperand(3), /* IsScalar */ true);
+    assert(Pivot->getType()->isIntegerTy() && "Pivot should be an integer.");
+    return Builder.CreateIntrinsic(Intrinsic::vp_merge, {OnTrue->getType()},
```
[the remark this answers is not in the document]
Comment (Mel-Chen) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:672]: I think the method of `VPWidenIntrinsicRecipe` is quite concise, and I appreciate it. Although I expressed a new perspective https://github.com/llvm/llvm-project/pull/109614#issuecomment-2396841914 that we don’t need to add an EVL recipe for every recipe, this doesn’t affect the use of `VPWidenIntrinsicRecipe` for the `vp.merge` use case.

The original reason of using `VPInstruction` was to avoid creating too many recipes. After all, I believe `merge until pivot` is not a commonly used operation.

In conclusion, switching to `VPWidenIntrinsicRecipe` is reasonable, especially if it can be reused in other places.

**Comment by Mel-Chen** [comment]
@fhahn Change to emit VPWidenIntrinsicRecipe. If you think this appoarch is better, please let me know. I will remove unused legacy code for VPInstruction::MergeUntilPivot later.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 4]
```diff
@@ -1125,6 +1125,8 @@ RISCVTTIImpl::getIntrinsicInstrCost(const IntrinsicCostAttributes &ICA,
         *FOp, ICA.getArgTypes()[0], UI->getPointerAlignment(),
         UI->getOperand(1)->getType()->getPointerAddressSpace(), CostKind);
   }
+  case Intrinsic::vp_merge:
+    return 1;
```
[the remark this answers is not in the document]
Comment (ElvisWang123) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1129]: Sure, I will implement the instruction cost of `vp_merge` in `getIntrinsicInstrCost()`.

[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1129]: Ideally this would be done separately from the patch, with the cost model tests independent of LV

[the remark this answers is not in the document]
Comment (Mel-Chen) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1129]: Yes, definitely.

[the remark this answers is not in the document]
Comment (ElvisWang123) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1129]: PR #112327 landed. We can get the instruction cost for `vp_merge` from `getintrinsicInstrCost()` in RISCVTTI.

**Comment by Mel-Chen** [comment]
Rebased, and updated commit title and log. Please take a look, thanks.

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 5]
```diff
@@ -9386,10 +9397,8 @@ void LoopVectorizationPlanner::adjustRecipesForReductions(
                cast<VPInstruction>(&U)->getOpcode() ==
                    VPInstruction::ComputeReductionResult;
       });
-      if (PreferPredicatedReductionSelect ||
-          TTI.preferPredicatedReductionSelect(
-              PhiR->getRecurrenceDescriptor().getOpcode(), PhiTy,
-              TargetTransformInfo::ReductionFlags()))
+      if (CM.usePredicatedReductionSelect(
+              PhiR->getRecurrenceDescriptor().getOpcode(), PhiTy))
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:9401]: Would it be sufficient to adjust the reduction phi recipe when introducing EVL recipes instead?

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp [hunk 6]
```diff
@@ -1392,7 +1395,19 @@ static void transformRecipestoEVLRecipes(VPlan &Plan, VPValue &EVL) {
                                                   TypeInfo.inferScalarType(Sel),
                                                   false, false, false);
               })
-
+              .Case<VPInstruction>([&](VPInstruction *VPI) -> VPRecipeBase * {
+                VPValue *LHS, *RHS;
+                if (!match(VPI, m_Select(m_Specific(HeaderMask), m_VPValue(LHS),
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:1400]: Would be good to document the code here

Do we need this for any select, or just the ones for out-loop reductions?

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp (hunk 5 above)
Comment (Mel-Chen) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:9465]: It can be adjusted in the EVL transformation. patch 8a3982fa6c65c08271b43194698d27e4cb29d0f4
But I don't recommend this. Such an implementation is more complicated and may lose the opportunity to sink the non-predicated reduction select outside the loop in the future VPlan transformation.
Could you point out why you want to adjust it in the EVL transformation?

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp (hunk 6 above)
Comment (Mel-Chen) [reply llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:1404]: doc it dc2da769fabe791feee128e4e401e55301c4881a
As far as I know, currently only out-loop reduction requires this transformation.

**Comment by Mel-Chen** [comment]
Rebase, drop redundant blank, and ping.

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp (hunk 6 above)
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:1503]: right, what I meant was that there could be other selects and we are not checking for reductions here?

File: llvm/lib/Transforms/Vectorize/VPlan.h [hunk 7]
```diff
@@ -1694,6 +1694,21 @@ class VPWidenIntrinsicRecipe : public VPRecipeWithIRFlags {
...
+        VectorIntrinsicID(VectorIntrinsicID), ResultTy(Ty) {
+    LLVMContext &Ctx = Ty->getContext();
+    AttributeList Attrs = Intrinsic::getAttributes(Ctx, VectorIntrinsicID);
+    MemoryEffects ME = Attrs.getMemoryEffects();
+    MayReadFromMemory = ME.onlyWritesMemory();
+    MayWriteToMemory = ME.onlyReadsMemory();
+    MayHaveSideEffects = MayWriteToMemory ||
+                         Attrs.hasFnAttr(Attribute::NoUnwind) ||
+                         !Attrs.hasFnAttr(Attribute::WillReturn);
+  }
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/VPlan.h:1710]: Can this replace the constructor above?

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 8]
```diff
@@ -1486,6 +1486,17 @@ class LoopVectorizationCostModel {
     return InLoopReductions.contains(Phi);
   }
 
+  /// Returns true if the predicated reduction select should be used.
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:1489]:
```suggestion
  /// Returns true if the predicated reduction select should be used to set the incoming value for the reduction phi.
```
?

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp (hunk 5 above)
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:9473]: At the moment, everything EVL related is applied during the transform that introduces EVL recipes; one potential issue is that we assume EVL is used here, but the transform may not apply.

Don't have any strong preferences, doing it later indeed seems to require some extra work

**Review by fhahn — COMMENTED** [review]
Thanks for the latest updates. Would be good for someone more familiar with EVL intrinsics to also take a look.

**Comment by alexey-bataev** [comment]
> Thanks for the latest updates. Would be good for someone more familiar with EVL intrinsics to also take a look.

Looks good to me

File: llvm/lib/Transforms/Vectorize/VPlan.h (hunk 7 above)
Comment (Mel-Chen) [reply llvm/lib/Transforms/Vectorize/VPlan.h]: Yes, separated to #113890

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp (hunk 8 above)
Comment (Mel-Chen) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp]: Formatted and committed it, thanks.

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp (hunk 6 above)
Comment (Mel-Chen) [reply llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:1503]: I understand your concern. However, I don’t think we need to perform a check specifically for reductions. As long as the `VPInstruction::select` matches the form `select(HeaderMask, LHS, RHS)`, it is correct to convert it to `vp.merge(all-true, LHS, RHS, EVL)`, whether or not it’s a predicated reduction select.

The difference between `vp.merge` and `vp.select` is that for all result lanes at positions greater or equal than EVL, `vp.select` sets them as  undefined, whereas `vp.merge` sets those lanes to `RHS` (the value on false). Thus, `VPWidenSelectRecipe` and `VPInstruction::select` **without** a header mask condition can be converted to `vp.select` because we don't care the results at inactive lanes, while `VPInstruction::select` **with** a header mask condition must be converted to `vp.merge`.

**Comment by Mel-Chen** [comment]
Ping

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp (hunk 6 above)
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:1503]: Sounds good, but then the comment should be updated to not refer to predicted reductions I think

Comment (Mel-Chen) [reply llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:1492]: 53d9ac6b0aa6d890e82cdffd53cd61faa025f772
Agree, replace `predicated reduction select` with `select with a header mask condition`.

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp (hunk 5 above)
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:9469]: Sounds good. to me

**Review by fhahn — APPROVED** [review]
LGTM, thanks!
